tendermint/src/serializers.rs

```diff
@@ -1,71 +1,199 @@
 //! Serde serializers
+//!
+//! Serializers and deserializers for a transparent developer experience.
+//!
+//! All serializers are presented in a serializers::<Rust_type_name>::<JSON_representation_name> format.
+//!
+//! This example shows how to serialize Vec<u8> into different types of strings:
+//! ```
+//! use serde::{Serialize, Deserialize};
+//! use tendermint::serializers;
+//!
+//! #[derive(Serialize, Deserialize)]
+//! struct ByteTypes {
+//!
+//!     #[serde(with="serializers::bytes::hexstring")]
+//!     hexbytes: Vec<u8>,
+//!
+//!     #[serde(with="serializers::bytes::base64string")]
+//!     base64bytes: Vec<u8>,
+//!
+//!     #[serde(with="serializers::bytes::string")]
+//!     bytes: Vec<u8>,
+//!
+//! }
+//! ```
+//!
+//! Available serializers:
+//! i64                  <-> string:               #[serde(with="serializers::primitives::string")]
+//! u64                  <-> string:               #[serde(with="serializers::primitives::string")]
+//! std::time::Dureation <-> nanoseconds as string #[serde(with="serializers::timeduration::string")]
+//! Vec<u8>              <-> HexString:            #[serde(with="serializers::bytes::hexstring")]
+//! Vec<u8>              <-> Base64String:         #[serde(with="serializers::bytes::base64string")]
+//! Vec<u8>              <-> String:               #[serde(with="serializers::bytes::string")]
+//!
+//! Any type that has the "FromStr" trait can be serialized into a string with serializers::primitives::string.
+//!
 
 use crate::account::{Id, LENGTH};
 use crate::{block, Hash, Signature};
 use serde::{de::Error as _, Deserialize, Deserializer, Serialize, Serializer};
 use std::str::FromStr;
-use std::time::Duration;
-use subtle_encoding::{base64, hex};
 
-/// Parse `i64` from a JSON string
-pub(crate) fn parse_i64<'de, D>(deserializer: D) -> Result<i64, D::Error>
-where
-    D: Deserializer<'de>,
-{
-    String::deserialize(deserializer)?
-        .parse::<i64>()
-        .map_err(|e| D::Error::custom(format!("{}", e)))
-}
+/// Serialize/deserialize primitive types (i64, u64, etc)
+pub mod primitives {
```

Comment:
Isn't `primitives` misleading? Although this is currently only used for primitive types, this is actually more general: `serializers::primitives::string` can be used for all types `T` that impl `FromStr` as far as I understand.

As this doesn't contain any submodules we could "flatten" this mod `serializers::string::from_str`?

Reply:
In a later step I updated the documentation to say "you can use this for anything that implements FromStr". So, you are right in your assumption.
I wanted to keep it embedded like this for convenience reasons for the developer: if you check the documentation, all built-in types are deserialized using "serializers::<Rust_nickname>::<JSON_object type>", or "serialize from A to B". If you flatten the modules (the modules don't really add any other additional value), it will work, but you lose this consistency.

I leave it up to you. If you say flatten it, I'll do it.

Reply:
hmm, that makes sense. For someone who usually does trust code more than documentation, primitives is still a bit misleading though (as you can use it on any type basically). If you really want to keep the module structure as is, then maybe `serializers::from_str::string` is still similar enough to: `"serializers::<Rust_nickname>::<JSON_object type>"``?

Reply:
To be honest, it should be something like `serializers::u64::string` and `serializers::i64::string` and so on, but that seemed overkill. Ok, I'll byte, let's rename.

Would you accept `serializers::from_str`? That kind of implies that you need a type that has a FromStr trait and it doesn't make any assumptions on what else that type is. It has one less module also. :)
And I'll rename the Time::Duration one to `serializers::timeduration` or `serializers::time_duration` or similar if you want.

The bytes thing still has multiple options to deserialize from. How about
`serializers::hexstring`
`serializers::base64string`
`seriliazers::bytestring` (that's the regular string to byte array serializer) ?

You'll have to implicitly know that you need to apply these to `Vec<u8>` but it's not bad if you can look at any example code.

Reply:
> Would you accept serializers::from_str? 
> serializers::time_duration 
> serializers::hexstring
> serializers::base64string
> seriliazers::bytestring 

Yes, sounds good to me!

Reply:
Uh, one last question: the step 3 PR based on this (well, on step 2, that's based on this step 1) breaks up the serializer.rs file into multiple files to support it better. If I do the rename here, I'm worried it won't go over well in that PR. (The curse of basing work on ongoing work.)

Option1: change it here, let's see the fallout and learn from it
Option2: we keep it as-is here, and the the change in the step 3 PR.

I'm fine doing it either way. What's your opinion?

Reply:
As it is a simple renaming I can't anticipate a bigger problem when moving these into separate files. As you will have to try it anyways probably better to go with option1 now while it is still a smaller change.

Reply:
I can implement it later this afternoon. Don’t wait for it today, sorry. We’ll finish merging this when you’re online next time.

Reply:
Thinking about it: `from_str` could also be just named `string` if you like (only one way it's `FromStr` the other it requires `fmt::Disply`). And if you want, we can keep the wrapping `bytes` module as it actually bundles something.

Reply:
Wait, I'll open a quick PR against this branch with the first steps to take this off your plate and get a feel for the implications.

Reply:
#264 (only string and timeduration)

Reply:
Thanks, that was awesome! I've reviewed it and merged.

Reply:
😊 Let's merge this and do any followup changes regarding the structure in the other step-PRs. My understanding is that some of the modules will change later anyways.